Add ARM MUSL targets

timonvo opened the pull request:
The targets are:
- `arm-unknown-linux-musleabi`
- `arm-unknown-linux-musleabihf`
- `armv7-unknown-linux-musleabihf`

These mirror the existing `gnueabi` targets.

All of these targets produce fully static binaries, similar to the
x86 MUSL targets.

For now these targets can only be used with `--rustbuild` builds, as
rust-lang/compiler-rt#22 only made the
necessary compiler-rt changes in the CMake configs, not the plain
GNU Make configs.

I've tested these targets GCC 5.3.0 compiled again musl-1.1.12
(downloaded from http://musl.codu.org/). An example `./configure`
invocation is:

```
./configure \
    --enable-rustbuild
    --target=arm-unknown-linux-musleabi \
    --musl-root="$MUSL_ROOT"
```

where `MUSL_ROOT` points to the `arm-linux-musleabi` prefix.
Usually that path will be of the form
`/foobar/arm-linux-musleabi/arm-linux-musleabi`.

Usually the cross-compile toolchain will live under
`/foobar/arm-linux-musleabi/bin`. That path should either by added
to your `PATH` variable, or you should add a section to your
`config.toml` as follows:

```
[target.arm-unknown-linux-musleabi]
cc = "/foobar/arm-linux-musleabi/bin/arm-linux-musleabi-gcc"
cxx = "/foobar/arm-linux-musleabi/bin/arm-linux-musleabi-g++"
```

As a prerequisite you'll also have to put a cross-compiled static
`libunwind.a` library in `$MUSL_ROOT/lib`. This is similar to [how
the x86_64 MUSL targets are built](https://doc.rust-lang.org/book/advanced-linking.html).

r? @alexcrichton

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented on src/bootstrap/build/compile.rs:
```diff
@@ -83,7 +83,7 @@ pub fn std_link(build: &Build,
 
     if target.contains("musl") &&
-       (target.contains("x86_64") || target.contains("i686")) {
+       (target.contains("x86_64") || target.contains("i686") || target.contains("arm")) {
```
Hm perhaps this condition could be reversed to say `if target.contains("musl") && !target.contains("mips")` ?

timonvo replied:
Done

alexcrichton commented on src/bootstrap/build/native.rs:
```diff
@@ -124,8 +124,14 @@ pub fn compiler_rt(build: &Build, target: &str) {
             "armhf"
+          } else {
+            "arm"
+          }
```
The spacing here looks like 2-space tabs instead of 4 perhaps?

timonvo replied:
Done.

alexcrichton commented on src/bootstrap/build/sanity.rs:
```diff
@@ -88,7 +88,9 @@ pub fn check(build: &mut Build) {
-        if target.contains("musl") && (target.contains("x86_64") || target.contains("i686")) {
+        if target.contains("musl") && (target.contains("x86_64") ||
+                                       target.contains("i686") ||
+                                       target.contains("arm")) {
```
Same `!target.contains("musl")` as above

timonvo replied:
Done

alexcrichton commented on src/compiler-rt:
```diff
@@ -1 +1 @@
-Subproject commit 57315f7e07d09b6f0341ebbcd50dded6c20d782f
+Subproject commit e3b7e473a467c3b8c54c413526818c87d38aebc2
```
This is fine by me, but as a heads up it's probably gonna make this PR pretty rough to land :(

timonvo replied:
It's the only way the new targets can actually work, so I'll have to deal with it...

alexcrichton commented on src/liballoc_jemalloc/build.rs:
```diff
@@ -46,7 +46,16 @@ fn main() {
+       // atomic operations" error. This is because no "-march" flag will be
+       // passed to GCC, and then GCC won't define the
+       // "__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4" macro that jemalloc needs to
+       // select an atomic operation implementation.
```
wut

timonvo replied:
Did I not make any sense? I can try to clarify it further.

alexcrichton replied:
Oh sorry, to expand, that comment from me typically means something along the lines of: "good lord build systems are hilariously complicated and are essentially a never-ending bugs where nothing works unless someone's testing it somewhere"

In other words, the comment here was excellent, I'm just surprised by the need for it at all!

alexcrichton commented on src/librustc_back/target/arm_unknown_linux_musleabi.rs:
```diff
@@ -0,0 +1,49 @@
+    base.position_independent_executables = false;
+
+    // Most of these settings are copied from the arm_unknown_linux_gnueabi
+    // target.
```
If you're feeling ambitious, perhaps this could become a standard among our musl targets? They could take the corresponding Linux target and then modify the same set of options to become "musl specific"

It's somewhat unfortunate to duplicate both the definition of the target and the musl configuration in multiple places :(. Not your fault or anything of course, so can always wait for a future PR.

timonvo replied:
I'll take a stab at that tonight.

timonvo replied:
I've extracted the common options to musl_base.rs.

alexcrichton commented on src/librustc_back/target/armv7_unknown_linux_musleabi.rs:
```diff
@@ -0,0 +1,84 @@
+    // To deal with these circular dependencies we just force the compiler to
+    // link everything as a group, not stripping anything out until everything
+    // is processed. The linker will still perform a pass to strip out object
+    // files but it won't do so until all objects/archives have been processed.
```
It's ok to leave this comment out pointing to the x86_64 musl comment.

timonvo replied:
Ah yes, I actually meant to remove this target completely. There's no armv7 gnueabi target either, only a gnueabihf one.

timonvo replied:
Removed now.

alexcrichton commented:
Thanks for the PR @timonvo, and nice!

Also cc @rust-lang/tools, curious on thoughts of the idea on using `target_env = "musleabi"` and `"musleabihf"`. On one hand it's unfortunate how musl-specific code now needs *three* different `target_env` arguments instead of one, although it's consistent with how we treat the target "triple" parsing. It's also consistent with the behavior of how we treat the glibc targets (e.g. `gnueabi` and `gnueabihf`), although I *highly* doubt anyone's relying on that distinction.

I don't quite know what the best way to expose this would be. We could perhaps add a `target_libc` or also perhaps repurpose `target_env`, but curious what others think

timonvo commented:
Hi Alex,

I think I've addressed all your comments. Is there anything else you'd like to see before this can land? I was thinking I'd squash the last two commits into the `fe4e327` commit once you give the OK that is can be submitted.

Regarding the `target_env` issue, as I mentioned in an earlier PR I think this is something that should maybe be resolved, but I don't think this is the right place for it now. As for what my preferred fix is: I think a new `target_libc` field may be the easiest fix. I think it would also be the most clear way of handling the problem, as for any #cfg based on that value it'd become immediately clear why the conditional is there.

alexcrichton commented:
Hm yeah I was thinking on waiting for a resolution of musleabi vs just musl, but I think you're right that it's probably best to duke out elsewhere. If you wanna do some squashes I'll r+. Thanks @timonvo!

timonvo commented:
Just did the squash of the last two commits.

alexcrichton commented on src/librustc_back/target/arm_unknown_linux_musleabi.rs:
```diff
@@ -0,0 +1,33 @@
+        // It's important we use "gnueabi" and not "musleabi" here. LLVM uses it
+        // to determine the calling convention and float ABI, and it doesn't
+        // support the "musleabi" value.
+        llvm_target: "arm-unknown-linux-gnueabi".to_string(),
```
Is it intentional to pass "gneuabi" to LLVM instead of "musleabi"?

timonvo replied:
Yes it is. I believe the comment above explains why :)

alexcrichton replied:
Wow I can totally read

alexcrichton commented:
Just one minor nit but otherwise r=me

alexcrichton commented:
@bors: r+ 483d80582cb5d9e1069175d19d8ad8778f3aa900

bors commented:
:hourglass: Testing commit 483d805 with merge 9147703...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/914)

alexcrichton commented:
@timonvo in https://github.com/rust-lang/rust/issues/33244 we decided that the targets should just use `musl` as the `target_env` instead of "musleabi" and "musleabihf", could you make that change as well?

timonvo commented:
Sure, I'll try to make that change. I'm not sure when I'll have time in next week or two, so this may be on hold for a while.

bors commented:
:umbrella: The latest upstream changes (presumably #33359) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
Closing due to inactivity, but feel free to resubmit with a rebase!

japaric commented:
@timonvo If you don't mind, I'm going to try to rebase this.
